compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -941,7 +948,9 @@ convertExpr env0 e = do
             let fldIndex = fromJust (elemIndex x vs)
             let fldName = fldNames !! fldIndex
             recTyp <- convertType env (varType bind)
-            pure $ ERecProj (fromTCon recTyp) fldName scrutinee' `ETmApp` EUnit
+            if envLfVersion env `supports` featureTypeSynonyms
```

Comment:
Another suggestion I have is to move (almost) all of these ```envLfVersion env `supports` featureTypeSynonyms``` checks to a set of related utility functions, something like

```Haskell
mkDictProj :: Env -> TypeConApp -> FieldName -> LF.Expr -> LF.Expr
mkDictProj env tcon fieldName
  if envLFVersion env `supports` featureTypeSynonyms
    then EStructProj fieldName
    else ERecProj tcon fieldName
 
mkDictCon :: Env -> TypeConApp -> [(String, LF.Expr)] -> Expr
mkDictcon env tcon =
  if envLFVersion env `supports` featureTypeSynonyms
    then EStructCon
    else ERecCon tcon   
```

etc